### PR TITLE
Add kubevirt provider

### LINKED ISSUE #418: Add a KubeVirt provider
In order to not require a third-party closed source virtualization solution we would like to add a KubeVirt provider.

*Acceptance criteria:*

* It is possible to use KubeVirt as a provider
* The Kubernetes cluster that hosts the KubeVirt does not have to be the same one that hosts the machine-controller
* KubeVirt is part of our e2e test suite

### PR DESCRIPTION
**Pull request by alvaroaleman**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #418 

**Special notes for your reviewer**:

TODOs:

* [x] Fix migrateUID
* [x] Samples
* [x] Docs

```release-note
Add experimental support for a [Kubevirt](https://kubevirt.io/) provider
```

### CODE REVIEW

File: cmd/controller/main.go [hunk 1]
```diff
@@ -146,14 +146,23 @@ type controllerRunOptions struct {
 }
 
 func main() {
-	flag.StringVar(&kubeconfig, "kubeconfig", "", "Path to a kubeconfig. Only required if out-of-cluster.")
-	flag.StringVar(&masterURL, "master", "", "The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
+	// This is also being registerd in kubevirt.io/kubevirt/pkg/kubecli/kubecli.go so
+	// we have to guard it
+	//TODO: Evaluate alternatives to importing the CLI. Generate our own client? Use a dynamic client?
```
Comment (mrIncompetent) [inline cmd/controller/main.go]: Can we push upstream for a change?

File: pkg/cloudprovider/provider/kubevirt/provider.go [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+	"github.com/kubermatic/machine-controller/pkg/providerconfig"
+
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/Azure/go-autorest/autorest/to"
```
Comment (mrIncompetent) [inline pkg/cloudprovider/provider/kubevirt/provider.go]: Can you create a refactoring issue to create a dedicated package for such pointer helper functions? More a "make me feel less dirty" refactoring

File: test/e2e/provisioning/all_e2e_test.go [hunk 3]
```diff
@@ -41,6 +42,21 @@ func TestInvalidObjectsGetRejected(t *testing.T) {
...
 
+func TestKubevirtProvisioningE2E(t *testing.T) {
+	t.Parallel()
+
+	kubevirtKubeconfig := os.Getenv("KUBEVIRT_E2E_TESTS_KUBECONFIG")
+	if kubevirtKubeconfig == "" {
+		t.Fatalf("Unable to run kubevirt tests, KUBEVIRT_E2E_TESTS_KUBECONFIG must be set")
+	}
+
+	// Provisioning coreos images via kubevirt does not work, needs investigation
```
Comment (mrIncompetent) [inline test/e2e/provisioning/all_e2e_test.go]: Please create an issue so we don't forget about it + mention it in the docs

**Comment by mrIncompetent** [comment]
Great PR!
But the documentation needs to state that we support KubeVirt (Except with CoreOS)

File: pkg/cloudprovider/provider/kubevirt/provider.go [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	virtualMachineInstance := &kubevirtv1.VirtualMachineInstance{}
+	virtualMachineInstance.Name = string(machine.UID)
+	virtualMachineInstance.Namespace = metav1.NamespaceSystem
+	virtualMachineInstance.Spec.Domain.CPU = &kubevirtv1.CPU{Cores: uint32(c.CPUs)}
```
[the remark this answers is not in the document]
Comment (alvaroaleman) [reply pkg/cloudprovider/provider/kubevirt/provider.go]: Yeah, in a single statement my autocompletion does not work ;) Will fix it up.

File: cmd/controller/main.go (hunk 1 above)
Comment (alvaroaleman) [reply cmd/controller/main.go]: Opened https://github.com/kubevirt/kubevirt/issues/1793

File: pkg/cloudprovider/provider/kubevirt/provider.go (hunk 2 above)
Comment (alvaroaleman) [reply pkg/cloudprovider/provider/kubevirt/provider.go]: I removed the dependency on that package from the kubevirt provider instead

File: pkg/cloudprovider/provider/kubevirt/provider.go [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+		return errors.New("CPUs must be 1 or greater")
+	}
+	if c.MemoryMiB < 512 {
+		return errors.New("MemoryMiB must be 512 or greater")
+	}
+	if _, err := parseResources(c.CPUs, c.MemoryMiB); err != nil {
+		return err
+	}
+	if pc.OperatingSystem == providerconfig.OperatingSystemCoreos {
+		return fmt.Errorf("CoreOS is not supported")
```
[the remark this answers is not in the document]
Comment (alvaroaleman) [reply pkg/cloudprovider/provider/kubevirt/provider.go]: I removed that part

File: test/e2e/provisioning/all_e2e_test.go (hunk 3 above)
Comment (alvaroaleman) [reply test/e2e/provisioning/all_e2e_test.go]: https://github.com/kubermatic/machine-controller/issues/422

**Comment by alvaroaleman** [comment]
@mrIncompetent All comments addressed, all tests green, PTAL

**Pull request merged**